### DISCUSSION TITLE
Refresh a certain Flet element in a for loop

### DISCUSSION THREAD

**Discussion by 16Integer** [flet-dev/flet#1725] [Q&A]
Hi, I'm currently creating a Password Manager program using Flet. I am trying to figure out how if I have a for loop for every 4 lines of a file to show text, with a remove button for each 4 lines, how I can make the remove button remove the container it was in, because at the moment to display the updated text, you have to rerun the application.

**Comment by FeodorFitsner** [comment]
Would you post a simple code demonstrating your issue?

**Reply by 16Integer** [reply]
Hi. Here's an example of what would occur in my actual code.

```
import flet as ft

def button_cmd(page, text):
    page.remove(text)


def main(page: ft.page):
    page.title = "Example GUI"
    page.horizontal_alignment = "center"
    page.vertical_alignment = "center"
    page.window_width = 420
    page.window_height = 420
    example_title = ft.Text(
        "My App", font_family="Arial, sans-serif", size=26, weight="bold"
    )
    example_text = ft.Text("This is my app example.")
    page.add(example_title, example_text)
    for i in range(5):
        text = ft.Text(f"Number {i}")
        example_button = ft.ElevatedButton(
            text="Click me!", on_click=lambda e: button_cmd(page, text)
        )
        page.add(text, example_button)
    page.update()


if __name__ == "__main__":
    ft.app(target=main)
```

As the text variable's last value will be whatever the value of the last time it was assigned in the _for loop_, clicking the button simply removes the last text that was created in the loop. Afterwards, clicking the button simply gives the error:

```
Exception in thread Thread-22 (<lambda>):
Traceback (most recent call last):
  File "C:\Python3.11.4\Lib\threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "C:\Python3.11.4\Lib\threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "C:\projects\tests\tests.py", line 28, in <lambda>
    text="Click me!", on_click=lambda e: button_cmd(page, text)
                                         ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\tests\tests.py", line 14, in button_cmd
    page.remove(text)
  File "C:\Python3.11.4\Lib\site-packages\flet_core\page.py", line 308, in remove
    self._controls.remove(control)
ValueError: list.remove(x): x not in list
```

Please let me know how I can make the button remove the appropriate text by clicking the button.

**Reply by ndonkoHenri** [reply]
I made some changes and seem to fixed your issue:
- ```python
    example_button = ft.ElevatedButton(text="Click me!", on_click=button_cmd, data=text)

The [`data`](https://flet.dev/docs/controls/#data) prop can be used to store anything you wish (of any data type). In this case I store a ref to the text.


- ```python
    def button_cmd(e):
        e.page.remove(e.control.data)

`e.control.data` gets the data(ref) stored as seen in the first point concerning `example_button `.
For the above the function, I will suggest you to use a `try:except` clause.

Hope it helped.

**Reply by 16Integer** [reply]
Hi. Thanks for this. However, how can I achieve this if I am calling the function with an argument as it gives the error that e is not defined?

**Reply by ndonkoHenri** [reply]
In your `lambda`, pass `e.control.data`, then pass it in the `remove()` of `button_cmd` to have it removed from the page.

**Reply by 16Integer** [reply]
Any way to do it with Columns and also remove multiple elements from the page?

**Reply by ndonkoHenri** [reply]
I don't seem to get your point. Please provide some more context and information.

**Reply by 16Integer** [reply]
As in, if I were packing the items into a column widget, for some reason, this method doesn't work. Also, by using this method, how can I remove multiple widgets instead of just one, ex: I have a button and text in a for loop being assigned to a column created outside the for loop (for scrolling purposes). I want to remove the correct button and text when the button is clicked.

**Reply by ndonkoHenri** [reply]
Does this match your needs?
```python
import flet as ft


def button_cmd(remove_from, controls):
    if controls:
        for c in controls:
            remove_from.controls.remove(c)
    remove_from.update()


def main(page: ft.page):
    page.title = "Example GUI"
    page.horizontal_alignment = "center"
    page.vertical_alignment = "center"
    page.window_width = 420
    page.window_height = 420

    example_title = ft.Text("My App", font_family="Arial, sans-serif", size=26, weight="bold")
    example_text = ft.Text("This is my app example.")
    example_col = ft.Column()
    
    page.add(example_title, example_text, example_col)

    for i in range(5):
        text = ft.Text(f"Title {i}")
        example_button = ft.ElevatedButton(
            text="Click me!",
            data=[text],    # this is a list; you can add some more controls here; it is unpacked below
            on_click=lambda e: button_cmd(
                remove_from=example_col,
                controls=[*e.control.data, e.control] # notice the unpack sign (*)
            )
        )
        example_col.controls.extend([text, example_button])
    page.update()


if __name__ == "__main__":
    ft.app(target=main)

```

**Reply by 16Integer** [reply]
Thank you so much! You've helped me quite a bit. I was stuck on this for a while.
